fcrepo/fcrepo
# Allow concurrent additions to a single parent resource

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3800

Creating as draft until we figure out https://github.com/fcrepo/fcrepo/pull/2019

# What does this Pull Request do?
* Adds a new ResourceLockType enum to differentiate between exclusive and non-exclusive locks.
* Adds a new ResourceLock interface and implementation which stores the transaction ID, resource ID and ResourceLockType
* Alters the Transaction interface and implementation to add a new lockResourceNonExclusive (the default is still an exclusive lock)
* Alters the InMemoryLockManager to allow multiple resources to hold a non-exclusive lock on the same resource, but does not allow a non-exclusive and exclusive lock to occur on the same resource at the same time (unless it is the same transaction, i.e. create and delete a resource in a single transaction.)

# How should this be tested?

1. Create an object 
     ```
     curl -v -XPUT -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/A
     ```
2. Create 1 long-running transaction
     ```
     curl -v -XPOST -ufedoraAdmin:fedoraAdmin -XPOST http://localhost:8080/rest/fcr:tx
     ```
3. Do a POST to the new object inside the transaction
     ```
     curl -v -XPOST -ufedoraAdmin:fedoraAdmin -XPOST -H"Atomic-ID: http://localhost:8080/rest/fcr:tx/<what ever your transaction ID is>" http://localhost:8080/rest/A
     ```
4. Do a similar POST without the long running transaction
     ```
     curl -v -XPOST -ufedoraAdmin:fedoraAdmin -XPOST http://localhost:8080/rest/A
     ```

With this PR step 4 succeeds, without you get a 409 Conflict and message like `Cannot update info:fedora/A because it is being updated by another transaction.`

If (with this PR) you commit your transaction
```
curl -v -XPUT -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/fcr:tx/<whatever your transaction ID is>
```
you will see two children for `http://localhost:8080/rest/A`

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
